File: chapter1/hash1.c
// simple hash function

#include "stdio.h"

#define hashsize(n) ((unsigned long)1 << (n))
#define hashmask(n) (hashsize(n) - 1)

unsigned long oaat(char *key, unsigned long len, unsigned long bits);

int main(void) {
  long snowflake[] = {1, 2, 3, 4, 5, 6};
  unsigned long code = oaat((char *)snowflake, sizeof(snowflake), 17);
  printf("%lu\n", code);
  printf("%ld\n", code % 100000);
  return 0;
}

unsigned long oaat(char *key, unsigned long len, unsigned long bits) {
  unsigned long hash, i;
  for (hash = 0,i = 0; i < len; i++) {
    hash += key[i];
    hash += (hash << 10);
    hash ^= (hash >> 6);
  }
  hash += (hash << 3);
  hash ^= (hash >> 11);
  hash += (hash << 15);
  return hash & hashmask(bits);
}

File: chapter1/hash1.rs
fn hashsize(n: u64) -> u64
{
    1_u64 << n
}


fn hashmask(n: u64) -> u64
{
    hashsize(n) - 1_u64
}

fn ooat(arr: &[u64],bits: u64) -> u64
{
    let mut hash: u64 = 0;
    
    for i in 0..arr.len()
    {
        hash += arr[i];
        hash += (hash << 10);
        hash ^= (hash >> 6);
    }
    
    hash += (hash << 3);
    hash ^= (hash >> 11);
    hash += (hash << 15);

    hash & hashmask(bits)
    
}

fn main()
{
   let snowflake = [1,2,3,4,5,6];
   let code = ooat(&snowflake,17);
   println!("{}",code);
   println!("{}",code % 100_000);
}